typescript/schedules.md

```diff
@@ -0,0 +1,1147 @@
+   */
+  interceptors?: ScheduleClientInterceptors;
+
+  /**
+   * Identity to report to the server
+   *
+   * @default `${process.pid}@${os.hostname()}`
+   */
+  identity?: string;
+
+  /**
+   * Connection to use to communicate with the server.
+   *
+   * By default `ScheduleClient` connects to localhost.
+   *
+   * Connections are expensive to construct and should be reused.
+   */
+  connection?: ConnectionLike;
+
+  /**
+   * Server namespace
+   *
+   * @default default
+   */
+  namespace?: string;
+}
+
+export type ScheduleClientOptionsWithDefaults = Replace<
+  Required<ScheduleClientOptions>,
+  {
+    connection?: ConnectionLike;
+  }
+>;
+export type LoadedScheduleClientOptions = ScheduleClientOptionsWithDefaults & {
+  loadedDataConverter: LoadedDataConverter;
+};
+
+export function defaultScheduleClientOptions(): ScheduleClientOptionsWithDefaults {
+  return {
+    dataConverter: {},
+    // The equivalent in Java is ManagementFactory.getRuntimeMXBean().getName()
+    identity: `${process.pid}@${os.hostname()}`,
+    interceptors: {},
+    namespace: 'default',
+  };
+}
+
+/**
+ * Policy for overlapping Actions.
+ */
+export enum ScheduleOverlapPolicy {
+  /**
+   * Use server default (currently SKIP).
+   *
+   * TODO remove this field if this issue is implemented: https://github.com/temporalio/temporal/issues/3240
+   */
+  UNSPECIFIED = 0,
+
+  /**
+   * Don't start a new Action.
+   */
+  SKIP,
+
+  /**
+   * Start another Action as soon as the current Action completes, but only buffer one Action in this way. If another
+   * Action is supposed to start, but one Action is running and one is already buffered, then only the buffered one will
+   * be started after the running Action finishes.
+   */
+  BUFFER_ONE,
+
+  /**
+   * Allows an unlimited number of Actions to buffer. They are started sequentially.
+   */
+  BUFFER_ALL,
+
+  /**
+   * Cancels the running Action, and then starts the new Action once the cancelled one completes.
+   */
+  CANCEL_OTHER,
+
+  /**
+   * Terminate the running Action and start the new Action immediately.
+   */
+  TERMINATE_OTHER,
+
+  /**
+   * Allow any number of Actions to start immediately.
+   *
+   * This is the only policy under which multiple Actions can run concurrently.
+   */
+  ALLOW_ALL,
+}
+
+checkExtendsWithoutPrefix<
+  temporal.api.enums.v1.schedule.ScheduleOverlapPolicy,
+  ScheduleOverlapPolicy,
+  'SCHEDULE_OVERLAP_POLICY_'
+>();
+
+export interface Backfill {
+  /** Time range to evaluate Schedule in. */
+  start: Date;
+  end: Date;
+
+  /**
+   * Override the Overlap Policy for this request.
+   */
+  overlap?: ScheduleOverlapPolicy;
+}
+
+/**
+ * The range of values depends on which, if any, fields are set:
+ *
+ * ```
+ * {} -> all values
+ * {start} -> start
+ * {start, end} -> every value from start to end (implies step = 1)
+ * {start, step} -> from start, by step (implies end is max for this field)
```

Comment:
sorry, this got a little confused. this description I wrote, that you copied in here, was a suggestion for how to unparse cron strings out of structured values. it is not the semantics of the new proto message Range. specifically: step defaults to 1, but end defaults to start. therefore {start:3, step:2} does not mean 3,5,7,..., it means only 3. you (either user or sdk) has to fill in end.

the actual code is basically:
```
  if (step == 0) step = 1;
  if (end < start) end = start;
  for (; start <= end; start += step)
    match(start);
```
you could implement these semantics by filling in `end` yourself but it might make more sense not to, for consistency